### PR TITLE
[Feature] SingleGap GPSR Higher order

### PR DESCRIPTION
**Pull request by dominikandreasseitz**
Enabing higher order gradient computation in PSRExpectation

### CODE REVIEW

File: pyqtorch/adjoint.py [hunk 1]
```diff
@@ -48,8 +48,8 @@ class AdjointExpectation(Function):
     def forward(
         ctx: Any,
         circuit: QuantumCircuit,
-        observable: Observable,
         state: Tensor,
+        observable: Observable,
```
Comment (Roland-djee) [inline pyqtorch/adjoint.py:52]: Why moving it here ?

File: pyqtorch/api.py [hunk 2]
```diff
@@ -92,11 +92,12 @@ def expectation(
         return inner_prod(state, observable.run(state, values)).real
     elif diff_mode == DiffMode.ADJOINT:
         return AdjointExpectation.apply(
-            circuit, observable, state, values.keys(), *values.values()
+            circuit, state, observable, values.keys(), *values.values()
```
Comment (Roland-djee) [inline pyqtorch/api.py:95]: Yeah, is there a reason for this ?

File: pyqtorch/gpsr.py [hunk 3]
```diff
@@ -92,56 +100,98 @@ def forward(
...
-        grads_dict = {k: None for k in values.keys()}
+        """The PSRExpectation backward call.
+
+        Note that only operations with two distinct eigenvalues
+        from their generator (i.e., compatible with single_gap_shift)
+        are supported at the moment.
+
+        Arguments:
+            ctx: Context object for accessing stored information.
+            grad_out: Current jacobian tensor.
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:111]:
```suggestion
            ctx (Any): Context object for accessing stored information.
            grad_out (Tensor): Current jacobian tensor.
```

File: pyqtorch/gpsr.py [hunk 4]
```diff
@@ -92,56 +100,98 @@ def forward(
...
+
+        Returns:
+            Updated jacobian tensor.
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:114]:
```suggestion
            A tuple of updated jacobian tensor.
```

File: pyqtorch/gpsr.py [hunk 5]
```diff
@@ -92,56 +100,98 @@ def forward(
...
+
+        Raises:
+            ValueError: When operation is not supported.
+        """
+
+        values = param_dict(ctx.param_names, ctx.saved_tensors)
         shift = torch.tensor(torch.pi) / 2.0
 
+        def expectation_fn(values: dict[str, Tensor]) -> torch.Tensor:
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:123]:
```suggestion
        def expectation_fn(values: dict[str, Tensor]) -> Tensor:
```

File: pyqtorch/gpsr.py [hunk 6]
```diff
@@ -92,56 +100,98 @@ def forward(
...
+                shift: Shift value. Defaults to torch.tensor(torch.pi)/2.0.
+
+            Returns:
+                Gradient evaluation for param_name.
+            """
+            shifted_values = values.copy()
+            shifted_values[param_name] = shifted_values[param_name] + shift
+            f_plus = expectation_fn(shifted_values)
+            shifted_values[param_name] = shifted_values[param_name] - 2 * shift
+            f_min = expectation_fn(shifted_values)
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:157]:
```suggestion
            f_minus = expectation_fn(shifted_values)
```

File: pyqtorch/gpsr.py [hunk 7]
```diff
@@ -92,56 +100,98 @@ def forward(
...
+            return (
+                spectral_gap
+                * (f_plus - f_min)
+                / (4 * torch.sin(spectral_gap * shift / 2))
+            )
+
+        def multi_gap_shift(*args, **kwargs) -> torch.Tensor:
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:164]:
```suggestion
        def multi_gap_shift(*args, **kwargs) -> Tensor:
```

File: pyqtorch/gpsr.py [hunk 8]
```diff
@@ -92,56 +100,98 @@ def forward(
...
+            """Implements multi gap PSR rule."""
+            raise NotImplementedError("Multi-gap is not yet supported.")
+
+        def vjp(operation: Parametric, values: dict[str, torch.Tensor]) -> torch.Tensor:
```
Comment (Roland-djee) [inline pyqtorch/gpsr.py:168]:
```suggestion
        def vjp(operation: Parametric, values: dict[str, Tensor]) -> Tensor:
```

File: pyqtorch/primitive.py [hunk 9]
```diff
@@ -102,6 +103,16 @@ def to(self, *args: Any, **kwargs: Any) -> Primitive:
         self._dtype = self.pauli.dtype
         return self
 
+    @cached_property
+    def eigenvals_generator(self) -> torch.Tensor:
```
Comment (Roland-djee) [inline pyqtorch/primitive.py:107]:
```suggestion
    def eigenvals_generator(self) -> Tensor:
```

File: pyqtorch/primitive.py [hunk 10]
```diff
@@ -102,6 +103,16 @@ def to(self, *args: Any, **kwargs: Any) -> Primitive:
...
+        return torch.linalg.eigvalsh(self.pauli).reshape(-1, 1)
+
+    @cached_property
+    def spectral_gap(self) -> torch.Tensor:
```
Comment (Roland-djee) [inline pyqtorch/primitive.py:111]:
```suggestion
    def spectral_gap(self) -> Tensor:
```

**Review by Roland-djee — COMMENTED** [review]
Thanks @chMoussa Looking good. Will approve after you've addressed my (minor) comments.

File: pyqtorch/adjoint.py (hunk 1 above)
Comment (chMoussa) [reply pyqtorch/adjoint.py:52]: To match both orders for adjoint and gpsr.

File: pyqtorch/api.py (hunk 2 above)
Comment (chMoussa) [reply pyqtorch/api.py:95]: We mixed the order at some point, now we made it consistent.

File: tests/test_differentiation.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+    cnot = pyq.CNOT(1, 2)
+    ops = [rx, cry, rz, cnot] * n_layers
+    circ = pyq.QuantumCircuit(n_qubits, ops)
+    obs = pyq.QuantumCircuit(n_qubits, [pyq.Z(0)])
+
+    theta_0_value = torch.pi / 2
+    theta_1_value = torch.pi
+    theta_2_value = torch.pi / 4
+
+    state = pyq.zero_state(n_qubits)
```
Comment (jpmoutinho) [inline tests/test_differentiation.py:30]: I would prefer `random_state`. I love me some random states.

File: tests/test_differentiation.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+    assert len(gradgrad_ad) == len(gradgrad_gpsr)
+
+    # check second order gradients
+    for i in range(len(gradgrad_ad)):
+        assert torch.allclose(gradgrad_ad[i], gradgrad_gpsr[i], atol=GRADCHECK_ATOL)
+
+
+@pytest.mark.xfail(raises=ValueError)
+@pytest.mark.parametrize("gate_type", ["scale", "hamevo", ""])
+def test_compatibility_gpsr(gate_type: str) -> None:
```
Comment (jpmoutinho) [inline tests/test_differentiation.py:255]: Does this test actually run like this? E.g. it "passes" only if a `ValueError` is raised?

I would typically write such a test with a `try: (...) except ValueError:` in the logic of the test itself, but if this accomplishes the same thing than I am fine with that.

**Review by jpmoutinho — APPROVED** [review]
Thanks @chMoussa @dominikandreasseitz 🙏 not much to say.

I think adding the same `check_support` logic for adjoint would be nice, but feel free to do that in a separate MR.

File: tests/test_differentiation.py (hunk 12 above)
Comment (chMoussa) [reply tests/test_differentiation.py]: It is similar but I will write it using `with pytest.raises(ValueError)` as a better syntax.

**Pull request merged**